src/transformStateWithClones.js

```diff
@@ -7,7 +7,38 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const statesArray = [];
+  const copyState = { ...state };
+
+  for (const action of actions) {
+    const { type, extraData, keysToRemove } = action;
+
+    switch (type) {
+      case 'addProperties':
+        Object.assign(copyState, extraData);
+        break;
+
+      case 'removeProperties':
+        for (const keyToRemove of keysToRemove) {
+          delete copyState[keyToRemove];
+        }
+        break;
+
+      case 'clear':
+        const keysToClear = Object.keys(copyState);
+
+        for (const keyToClear of keysToClear) {
```

Comment:
I believe `for...in` loop will do better here, for making it more simple. And we won't need Object.values in that case